src/window_manipulation.rs

```diff
@@ -0,0 +1,374 @@
+	/// [`cursor_y()`](WindowManipulation::cursor_y)
+	pub fn cursor_y(self) -> i16 {
+		self.cursor_pos().1
+	}
+
+	/// The difference between the given cursor position and the starting
+	/// [`cursor_pos()`](WindowManipulation::cursor_pos).
+	///
+	/// See also: [`diff_x(x: i16)`], [`diff_y(y: i16)`]
+	pub fn diff(self, cursor_pos: (i16, i16)) -> (i16, i16) {
+		(self.diff_x(cursor_pos.0), self.diff_y(cursor_pos.1))
+	}
+
+	/// The difference between the given x coordinate and the starting
+	/// [`cursor_x()`](WindowManipulation::cursor_x).
+	///
+	/// See also: [`diff(cursor_pos: (i16, i16))`](WindowManipulation::diff),
+	/// [`diff_y(y: i16)`](WindowManipulation::diff_y)
+	pub fn diff_x(self, x: i16) -> i16 {
+		x - self.cursor_pos().0
+	}
+
+	/// The difference between the given y coordinate and the starting
+	/// [`cursor_y()`](WindowManipulation::cursor_y).
+	///
+	/// See also: [`diff(cursor_pos: (i16, i16))`](WindowManipulation::diff),
+	/// [`diff_x(x: i16)`](WindowManipulation::diff_x)
+	pub fn diff_y(self, y: i16) -> i16 {
+		y - self.cursor_pos().1
+	}
+
+	/// Applies the window manipulation to the given window.
+	pub fn apply(self, conn: &Connection, cursor_pos: (i16, i16)) -> xcb::Result<()> {
+		let diff_x = self.diff_x(cursor_pos.0);
+		let diff_y = self.diff_y(cursor_pos.1);
+
+		match self {
+			Self::Moving {
+				window,
+				orig_coords,
+				..
+			} => {
+				set_position(
+					conn,
+					window,
+					(
+						orig_coords.0 as i32 + diff_x as i32,
+						orig_coords.1 as i32 + diff_y as i32,
+					),
+				);
+			}
+			Self::Resizing {
+				window, orig_size, ..
+			} => {
+				// Since the difference is a signed integer but the window dimensions are
+				// unsigned integers, we must apply the difference to the dimensions carefully.
+				let dimensions = (
+					// x
+					if diff_x > 0 {
+						// If `diff_x` is positive, we can just add the numbers.
+						orig_size.0 as u32 + diff_x as u32
+					} else {
+						// If `diff_x` is negative, we need to make sure it won't shrink the width
+						// too far.
+						if (-diff_x as u16) < orig_size.0 {
+							// If `diff_x` decreases the width by less than the width itself, we
+							// can subtract it.
+							orig_size.0 as u32 - -diff_x as u32
+						} else {
+							// If `diff_x` shrinks the width too much, we'll set the width to 1.
+							1
+						}
+					},
+					// y
+					if diff_y > 0 {
+						// If `diff_y` is positive, we can just add the numbers.
+						orig_size.1 as u32 + diff_y as u32
+					} else {
+						// If `diff_y` is negative, we need to make sure it won't shrink the
+						// height too far.
+						if (-diff_y as u16) < orig_size.1 {
+							// If `diff_y` decreases the height by less than the height itself, we
+							// can substract it.
+							orig_size.1 as u32 - -diff_y as u32
+						} else {
+							// If `diff_y` shrinks the height too much, we'll set the height to 1.
+							1
+						}
+					},
+				);
```

Comment:
Is there a better way to do this (like a builtin way or something)? If not, should this be moved to a separate utility function, or is it fine where it is?